xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/internal/doc/XWikiAttachmentList.java

```diff
@@ -0,0 +1,238 @@
+/*
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package com.xpn.xwiki.internal.doc;
+
+import java.util.ArrayList;
+
+import java.util.Collection;
+import java.util.ListIterator;
+import java.util.List;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.concurrent.ConcurrentSkipListMap;
+
+import com.xpn.xwiki.doc.XWikiAttachment;
+import com.xpn.xwiki.doc.XWikiDocument;
+
+import com.xpn.xwiki.internal.AbstractNotifyOnUpdateList;
+
+/**
+ * AttachmentList that holds elements in order of filename.
+ * 
+ * @version $Id$
+ * @param <E>
+ * @since 10.0RC1
+ */
+public class XWikiAttachmentList extends AbstractNotifyOnUpdateList<XWikiAttachment>
```

Comment:
Now that I walk your code more, because of the clear()/addAll() logic you actually seems to reimplement the whole  AbstractNotifyOnUpdateList write logic anyway, right ? So looks like there is no point in extending it after all, sorry for the mislead.

Another note: instead of clear/addAll, creating a new ArrayList would improve Thread safety.

So here is a proposal: extend org.apache.commons.collections4.list.AbstractListDecorator from which you override all the write methods to modify the Map instead and then re-create and change the decorated List.

Here is what I would have done from when starting from scratch: make XWikiDocument#attachmentList a simple ConcurrentSkipListMap and create a new XWikiAttachmentList-like each time getAttachmentList() is called (i.e. consider it like some retro compatibility instead of being the core the XWikiDocument attachment handling). But your version is OK also so it's just a "for your information" note :)

Reply:
@tmortagne why not just extend abstractnotifyonupdatelist? We get the same additional list methods and also have the onUpdate() and added(). Is there any added benefit to extending the listdecorator

Reply:
@tmortagne also just wondering, would you have returned an independent list from getAttachmentList() that just mirrors the map?  how would that handle new things being added to the list?

Reply:
I fixed all the minor changes that you asked, I'm not sure why the decorator is needed as we do reuse a few of the notifier methods and they automatically call update and added. I left it extending the notifier for now. The only limitation I see is that the list in the notifier can't be final since we want to set it to a new list instead of clear/addAll. I will be sleeping now, will check in in the morning! Let me know your thoughts!

Reply:
> @tmortagne also just wondering, would you have returned an independent list from getAttachmentList() that just mirrors the map? how would that handle new things being added to the list?

As I said it would be something similar to XWikiAttachmentList but only for getAttachmentList() so it would handle new things added to the Map. What getAttachmentList() is a very bad practice to I would have deprecated it and considered all this as a retro compatibility thing and introduce a new method which return an unmodifiable list. Basically the idea would be to not "pollute" the whole XWikiDocument with this.

> @tmortagne why not just extend abstractnotifyonupdatelist? We get the same additional list methods and also have the onUpdate() and added(). Is there any added benefit to extending the listdecorator

I much preference extending a more or less standard org.apache.commons.collections4.list.AbstractListDecorator when all I need is a decorator. You don't really need onUpdate() here since you override all the write methods (you have to since you can't rely on abstractnotifyonupdatelist implementation because you need to recreate the list from the Map).

Reply:
@tmortagne when you said "you don't need onUpdate()", did you mean that I don't need it at all or its a functionality of the notifier that can be easily rewritten?

Reply:
Also, listdecorator won't actually allow me to access the list right? So I will have to use clear/addAll?

Reply:
> Also, listdecorator won't actually allow me to access the list right? So I will have to use clear/addAll?

You can access it but even so if you recreate it from the map values every time all you need is to call #setCollection with the new ArrayList.

Reply:
It also will require a lot of casting, since the methods don't override if the parameter is an XWikiAttachment while it has to be an Object

Reply:
> @tmortagne when you said "you don't need onUpdate()", did you mean that I don't need it at all or its a functionality of the notifier that can be easily rewritten?

The only point of onUpdate() as defined by AbstractNotifyOnUpdateList is to be overwritten by an extending class to know when the list is modified but your are not writing a class to extend here. You don't really care about this generic feature, all you care about is updating the XWikiDocument medatada dirty flag. Also you are not "using" AbstractNotifyOnUpdateList onUpdate() feature since you are the one calling that method (since you reimplement all the writing methods). In short my point is that you don't really reuse from AbstractNotifyOnUpdateList more that you would get from AbstractListDecorator and I usually prefer to reuse Apache commons stuff than maintaining custom XWiki stuff.

Reply:
> It also will require a lot of casting, since the methods don't override if the parameter is an XWikiAttachment while it has to be an Object

I don't understand what you mean here.

Reply:
@tmortagne The casting thing was a misunderstanding on my part, I apologize. You said that I can access the list, but I don't know how we could do that as it would be a private field, right? In order to reassign it to a new list, we can't just get it from the decorate() (gets the list) method.

Reply:
> In order to reassign it to a new list, we can't just get it from the decorate() (gets the list) method.

You missed the following: "all you need is to call #setCollection with the new ArrayList" :)

Reply:
I did see that, but I don’t see how I can create a “new ArrayList”

Reply:
I don't understand, you are doing it in updateMap().

Reply:
@tmortagne Oh I see! I thought you wanted me to create a new list and then call setCollection on it but I don't think I can access the list field.

Reply:
In practice I don't think you need to access the list field for any other reason than reseting it.

Reply:
The setCollection method is a protected method part of AbstractCollectionDecorator which is extended by AbstractListDecorator and java does not allow super.super, how should I go about calling it?